File: DynamoDBTutorialNode/test/user/UserRepositoryTest.js
/**
 * http://usejsdoc.org/
 */

var UserRepository = require('../../user/UserRepository'),
	TableCreator = require('../../TableCreator'),
	async = require('async'),
	assert = require('assert');

describe('User Repository Test',function(){
    
	beforeEach(function(done) {
		TableCreator.deleteUsers(function(err,result) {
			
			if(err) {
				test.done(err);
				return;
			}
			
			
			TableCreator.createUsers(function(err,result) {
				
				if(err) {
					test.done(err);
					return;
				}
				
				UserRepository.insertUser("john@doe.com", "John Doe", new Date(), function(err,result) {
					
					if(err) {
						done(err);
					} else {
						done();
					}
				});

			});			
		});
	});
	
	it('Query user',function(done) {
		UserRepository.getUser("john@doe.com",function(err,result) {
			if(err) {
				done(err);
			} else {
				console.log(result);
				done();
			}
		});
	});
	
	it('Update user',function(done) {
		
		UserRepository.updateName("john@doe.com","Different Name",function(err,result) {
			if(err) {
				done(err);
			} else {
				UserRepository.getUser("john@doe.com",function(err,result) {
					if(err) {
						done(err);
					} else {
						assert.equal(result.Items[0].fullname,"Different Name");
						done();
					}
				});
			}	
		});
	});
	
	it('Update user conditionally',function(done) {
		
		UserRepository.updateConditionally("john@doe.com","Different Name","John",function(err,result) {
			if(err) {
				done(err);
			} else {
				UserRepository.getUser("john@doe.com",function(err,result) {
					if(err) {
						done(err);
					} else {
						assert.equal(result.Items[0].fullname,"Different Name");
						done();
					}
				});
			}	
		});
	});
	
	it('Add and increse counter',function(done){
		
		async.waterfall(
				[
				 function(callback) {
					UserRepository.addUpdateCounter("john@doe.com",function(err,result){
						if(err) {
							callback(err);
						} else {
							callback();
						}
					});
				 },
				 function(callback) {
					UserRepository.updateAndIncreaseCounter("john@doe.com","New Name",function(err,result){
						if(err) {
							callback(err);
						} else {
							callback();
						}
					});
				 },
				 function(callback) {
				    UserRepository.getUser("john@doe.com",callback);
				 }
						     
					 ],
		function(err,result){
			if(err) {
				done(err);
			} else {
				assert.equal(result.Items[0].fullname,"New Name");
				assert.equal(result.Items[0].counter,"1");
				
				done();
			}
		});
	});
	
	it("Get Register Date",function(done) {
		UserRepository.getRegisterDate("john@doe.com",function(err,result) {
			if(err) {
				done(err);
			} else {
				console.log(result);
				done();
			}
		});
	});
	
})

File: DynamoDBTutorial/src/test/java/com/gkatzioura/dynamodb/user/UserRepositoryTest.java
package com.gkatzioura.dynamodb.user;

import com.amazonaws.services.dynamodbv2.AmazonDynamoDB;
import com.amazonaws.services.dynamodbv2.AmazonDynamoDBClient;
import com.amazonaws.services.dynamodbv2.model.AttributeValue;
import com.amazonaws.services.dynamodbv2.model.ConditionalCheckFailedException;
import com.gkatzioura.dynamodb.TableCreator;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;

import java.util.Date;
import java.util.Map;

/**
 * Created by gkatzioura on 7/2/16.
 */
public class UserRepositoryTest {

    private AmazonDynamoDB amazonDynamoDB;
    private UserRepository userRepository;

    @Before
    public void setUp() {
        AmazonDynamoDB amazonDynamoDB = new AmazonDynamoDBClient();
        amazonDynamoDB.setEndpoint("http://localhost:8000");
        this.amazonDynamoDB = amazonDynamoDB;
        this.userRepository = new UserRepository(amazonDynamoDB);
        TableCreator tableCreator = new TableCreator(amazonDynamoDB);
        tableCreator.deleteUsersTable();
        tableCreator.createUserTable();
    }

    @Test
    public void testGetUser() {

        userRepository.insertUser("me@test.com","Mr mean",new Date());
        Map<String,AttributeValue> user = userRepository.getUser("me@test.com");

        Assert.assertEquals(user.get("fullname").getS(),"Mr mean");
    }

    @Test
    public void testGetRegisterDate() {

        userRepository.insertUser("me@test.com","Mr mean",new Date());
        Map<String,AttributeValue> user = userRepository.getRegisterDate("me@test.com");

        System.out.println("d");
    }

    @Test
    public void updateUser() {

        userRepository.insertUser("me@test.com","Mr mean",new Date());

        Map<String,AttributeValue> user = userRepository.getUser("me@test.com");

        userRepository.updateName("me@test.com","John Doe");

        Map<String,AttributeValue> updatedUser = userRepository.getUser("me@test.com");

        Assert.assertNotEquals(user.get("fullname").getS(),updatedUser.get("fullname").getS());
    }

    @Test
    public void updateUserConditionally() throws ConditionalCheckFailedException {

        userRepository.insertUser("me@test.com","Mr mean",new Date());

        Map<String,AttributeValue> user = userRepository.getUser("me@test.com");

        userRepository.updateConditionallyWithExpression("me@test.com","John Doe","Mr");

        Map<String,AttributeValue> updatedUser = userRepository.getUser("me@test.com");

        Assert.assertNotEquals(user.get("fullname").getS(),updatedUser.get("fullname").getS());
    }

    @Test
    public void testUpdateConditionallyWithAttributeEntries() throws ConditionalCheckFailedException {

        userRepository.insertUser("me@test.com","Mr mean",new Date());

        Map<String,AttributeValue> user = userRepository.getUser("me@test.com");

        userRepository.updateConditionallyWithAttributeEntries("me@test.com","John Doe","Mar");

        Map<String,AttributeValue> updatedUser = userRepository.getUser("me@test.com");

        Assert.assertNotEquals(user.get("fullname").getS(),updatedUser.get("fullname").getS());
    }

    @Test
    public void createAtomicCounter() {

        userRepository.insertUser("me@test.com","Mr mean",new Date());

        Map<String,AttributeValue> user = userRepository.getUser("me@test.com");

        userRepository.addUpdateCounter("me@test.com");

        Map<String,AttributeValue> userWithCounter = userRepository.getUser("me@test.com");

        userRepository.updateAndIncreaseCounter("me@test.com","John doe");

        Map<String,AttributeValue> counterIncreased = userRepository.getUser("me@test.com");

        Assert.assertEquals(userWithCounter.get("counter").getN(),"0");
        Assert.assertEquals(counterIncreased.get("counter").getN(),"1");

    }

}
